### PR TITLE
DPO Llava 1.5 and PaliGemma support

### LINKED ISSUE #1784: Supports of SFTTrainer / PPOTrainer / DPOTrainer for LLaVA-alike model
TRL SFTTrainer supports LLaVA (Large Language and Vision Assistant) as described in the following link [Vision Language Models Explained](https://huggingface.co/blog/vlms)

Is there any plan to release PPOTrainer and DPOTrainer for LLaVA? If not, could someone explain the concerns about implementing those trainers or suggest any alternatives? Thanks!

### PR DESCRIPTION
**Pull request by qgallouedec**
Closes #1784

Llava:

```sh
accelerate launch examples/scripts/dpo_visual.py \
    --dataset_name HuggingFaceH4/rlaif-v_formatted \
    --model_name_or_path llava-hf/llava-1.5-7b-hf \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 8 \
    --output_dir dpo_llava_rlaif-v \
    --bf16 \
    --torch_dtype bfloat16 \
    --use_peft \
    --lora_target_modules=all-linear
```

PaliGemma:

```sh
accelerate launch examples/scripts/dpo_visual.py \
    --dataset_name HuggingFaceH4/rlaif-v_formatted \
    --model_name_or_path google/paligemma-3b-pt-224 \
    --trust_remote_code \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 8 \
    --output_dir dpo_paligemma_rlaif-v \
    --bf16 \
    --torch_dtype bfloat16
```

### CODE REVIEW

**Review by kashif — APPROVED** [review]

File: examples/scripts/dpo_visual.py [hunk 1]
```diff
@@ -16,12 +16,13 @@
...
     --model_name_or_path HuggingFaceM4/idefics2-8b \
-    --per_device_train_batch_size 1 \
-    --gradient_accumulation_steps 16 \
+    --per_device_train_batch_size 2 \
+    --gradient_accumulation_steps 32 \
     --dataset_num_proc 32 \
     --output_dir dpo_idefics_rlaif-v \
     --bf16 \
     --torch_dtype bfloat16 \
+    --gradient_checkpointing \
```
Comment (qgallouedec) [inline examples/scripts/dpo_visual.py:25]: Allows the training to fit in the memory without the need to resize images. (resizing has been removed as well, see under)

File: examples/scripts/dpo_visual.py [hunk 2]
```diff
@@ -82,21 +83,40 @@
 
     model_kwargs = dict(
         revision=model_config.model_revision,
-        trust_remote_code=model_config.trust_remote_code,
         attn_implementation=model_config.attn_implementation,
         torch_dtype=torch_dtype,
-        use_cache=False if training_args.gradient_checkpointing else True,
```
Comment (qgallouedec) [inline examples/scripts/dpo_visual.py:88]: PaliGemma doesn't take `use_cache` as init argument

File: examples/scripts/dpo_visual.py [hunk 3]
```diff
@@ -82,21 +83,40 @@
 
     model_kwargs = dict(
         revision=model_config.model_revision,
-        trust_remote_code=model_config.trust_remote_code,
```
Comment (qgallouedec) [inline examples/scripts/dpo_visual.py:85]: See https://github.com/huggingface/trl/pull/1806

File: examples/scripts/dpo_visual.py [hunk 4]
```diff
@@ -82,21 +83,40 @@
...
+        pass  # the processor already has a valid chat template
+    elif model.config.model_type == "paligemma":
+        processor.chat_template = """{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}{% for message in messages %}<|im_start|>{% if message['role'] == 'user' %}USER: {% else %}ASSISTANT: {% endif %}{% for item in message['content'] if item['type'] == 'text' %}{{ item['text'] }}<|im_end|>{% endfor %}{% if message['role'] == 'user' %} {% else %}{{eos_token}}{% endif %}{% endfor %}{% if add_generation_prompt %}ASSISTANT: {% endif %}"""
+    elif model.config.model_type == "llava":
+        processor.chat_template = """{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}{% for message in messages %}{% if message['role'] == 'user' %}USER: {% else %}ASSISTANT: {% endif %}{% for item in message['content'] %}{% if item['type'] == 'text' %}{{ item['text'] }}{% elif item['type'] == 'image' %}<image>{% endif %}{% endfor %}{% if message['role'] == 'user' %} {% else %}{{eos_token}}{% endif %}{% endfor %}{% if add_generation_prompt %}ASSISTANT: {% endif %}"""
```
Comment (qgallouedec) [inline examples/scripts/dpo_visual.py:118]: We expect the chat templating to support this structure of data:

```
[{"role": "user", "content": [{"type": "image"}, {"type": "text", "text": "foo"}]}]
```

but idefics2 is the only model to support it. So we need to adapt the chat template of the other models.

File: examples/scripts/dpo_visual.py [hunk 5]
```diff
@@ -124,27 +144,9 @@
...
-        #        "chosen": [{"content": "What does detox mean?", "role": "user"}, {"content": "It means to get rid of the toxins.", "role": "assistant"}],
-        #        "rejected": [{"content": "What does detox mean?", "role": "assistant"}, {"content": "I don't know.", "role": "user"}]}
-        # Example 2:
-        # row = {"prompt": [{"content": "What does detox mean?", "role": "user"}],
-        #        "chosen": [{"content": "It means to get rid of the toxins.", "role": "assistant"}],
-        #        "rejected": [{"content": "I don't know.", "role": "user"}]}
-        if "prompt" in row and isinstance(row["prompt"], list):
-            row["prompt"] = processor.apply_chat_template(row["prompt"], tokenize=False)
-
+        row["prompt"] = processor.apply_chat_template(row["prompt"], tokenize=False)
```
Comment (qgallouedec) [inline examples/scripts/dpo_visual.py:147]: It turns out that this isn't the case for visual data: we always have a prompt. The reason is that it's the prompt that contains the image (the image can't be part of the output).

File: trl/trainer/dpo_trainer.py [hunk 6]
```diff
@@ -723,9 +723,18 @@ def build_tokenized_answer(self, prompt, answer, images=None):
         if self.is_vision_model:
             if answer.count("<image>") > 0:
                 raise NotImplementedError("Answer contains <image> token, which is not supported yet.")
-            full_tokenized = self.processor(prompt + answer, images=images, add_special_tokens=False)
+            if "add_special_tokens" in inspect.signature(self.processor).parameters:
+                processor_kwargs = {"add_special_tokens": False}
+            else:
+                processor_kwargs = {}
+            full_tokenized = self.processor(prompt + answer, images=images, **processor_kwargs)
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:730]: `add_special_token` is not an argument in all processors. So the only way is to check the signature.

File: trl/trainer/dpo_trainer.py [hunk 7]
```diff
@@ -762,22 +771,18 @@ def build_tokenized_answer(self, prompt, answer, images=None):
...
-                prompt_input_ids=prompt_input_ids,
-                prompt_attention_mask=prompt_attention_mask,
-                input_ids=answer_input_ids,
-                attention_mask=answer_attention_mask,
-            )
+            return_dict["prompt_pixel_values"] = full_tokenized["pixel_values"]
+        if "pixel_attention_mask" in full_tokenized:
+            return_dict["prompt_pixel_attention_mask"] = full_tokenized["pixel_attention_mask"]
+
+        return return_dict
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:785]: Equivalent, but simpler. The only motif is that it allows `full_tokenized` not to have an entry `"pixel_attention_mask"` (which is the case for Llava)

File: trl/trainer/dpo_trainer.py [hunk 8]
```diff
@@ -1037,10 +1049,13 @@ def concatenated_inputs(
...
-            concatenated_batch["pixel_values"] = batch["prompt_pixel_values"].repeat(2, 1, 1, 1, 1).to(device=device)
-            concatenated_batch["pixel_attention_mask"] = (
-                batch["prompt_pixel_attention_mask"].repeat(2, 1, 1, 1).to(device=device)
+            concatenated_batch["pixel_values"] = torch.cat(
+                [batch["prompt_pixel_values"], batch["prompt_pixel_values"]], dim=0
             )
+            if "prompt_pixel_attention_mask" in batch:
+                concatenated_batch["pixel_attention_mask"] = torch.cat(
+                    [batch["prompt_pixel_attention_mask"], batch["prompt_pixel_attention_mask"]], dim=0
+                )
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1058]: depending on the processor, the `prompt_pixel_values` can have a shape `(B, N, 3, W, H)` (idefics2) or `(B, 3, W, H)` (Llava). To make these lines compatible with both cases, we use self concatenation cat in the first dim instead of `repeat`.

Comment (kashif) [reply trl/trainer/dpo_trainer.py:1058]: you mean `dim=1`?

**Review by kashif — APPROVED** [review]

File: trl/trainer/dpo_trainer.py (hunk 8 above)
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:1058]: At this point, `pixel_values` is either 
- a batch of images (llava case), so shape `(B, C, W, H)`, or 
- a batch of list of images (idefics2 case), so shape `(B, N, C, W, H)`

The image is shared between chosen and rejected. So we need to duplicate the batch in the batch dimension. You want to end up with a shape  `(2*B, ...)`
- first case: `(2*B, C, W, H)`
- second case: `(2*B, N, C, W, H)`

The problem is that with `torch.repeat`, you need to pass the number of repetition in each dimension. So having something like

```python
if len(image.shape) == 4:
    pixel_values = prompt_pixel_values.repeat(2, 1, 1, 1)
elif len(image.shape) == 5:
    pixel_values  = prompt_pixel_values.repeat(2, 1, 1, 1, 1)
```

An easier solution is to write:

```python
pixel_values = torch.cat([prompt_pixel_values, prompt_pixel_values], dim=0)
```

File: examples/scripts/dpo_visual.py [hunk 9]
```diff
@@ -124,27 +144,9 @@
...
         row["chosen"] = processor.apply_chat_template(row["chosen"], tokenize=False)
         row["rejected"] = processor.apply_chat_template(row["rejected"], tokenize=False)
-
-        if "images" in row:
-            for img in row["images"]:  # Resize each image so the largest side is 640 pixels
-                img.thumbnail((640, 640))  # Resize the image to at most 640x640 pixels
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply examples/scripts/dpo_visual.py:147]: yep, we shouldn't do it or else it breaks

**Comment by kashif** [comment]
the tests are failing due to some time-outs i believe

**Comment by qgallouedec** [comment]
Re-running solved the issue

File: examples/scripts/dpo_visual.py (hunk 4 above)
Comment (zucchini-nlp) [reply examples/scripts/dpo_visual.py:118]: Btw, llava models have different chat templates depending on which llm backbone is used. We'll add chat templates directly to processor, but not very soon probably. 

Wondering if we need to support different templates here for each checkpoint

Comment (qgallouedec) [reply examples/scripts/dpo_visual.py:118]: We can definitely work on its support, yes.

**Comment by Liuziyu77** [comment]
> Closes #1784
> 
> Llava:
> 
> ```shell
> accelerate launch examples/scripts/dpo_visual.py \
>     --dataset_name HuggingFaceH4/rlaif-v_formatted \
>     --model_name_or_path llava-hf/llava-1.5-7b-hf \
>     --per_device_train_batch_size 1 \
>     --gradient_accumulation_steps 8 \
>     --output_dir dpo_llava_rlaif-v \
>     --bf16 \
>     --torch_dtype bfloat16 \
>     --use_peft \
>     --lora_target_modules=all-linear
> ```
> 
> PaliGemma:
> 
> ```shell
> accelerate launch examples/scripts/dpo_visual.py \
>     --dataset_name HuggingFaceH4/rlaif-v_formatted \
>     --model_name_or_path google/paligemma-3b-pt-224 \
>     --trust_remote_code \
>     --per_device_train_batch_size 1 \
>     --gradient_accumulation_steps 8 \
>     --output_dir dpo_paligemma_rlaif-v \
>     --bf16 \
>     --torch_dtype bfloat16
> ```

how to set other parameters, like learning rate?

**Comment by Liuziyu77** [comment]
> Closes #1784
> 
> Llava:
> 
> ```shell
> accelerate launch examples/scripts/dpo_visual.py \
>     --dataset_name HuggingFaceH4/rlaif-v_formatted \
>     --model_name_or_path llava-hf/llava-1.5-7b-hf \
>     --per_device_train_batch_size 1 \
>     --gradient_accumulation_steps 8 \
>     --output_dir dpo_llava_rlaif-v \
>     --bf16 \
>     --torch_dtype bfloat16 \
>     --use_peft \
>     --lora_target_modules=all-linear
> ```
> 
> PaliGemma:
> 
> ```shell
> accelerate launch examples/scripts/dpo_visual.py \
>     --dataset_name HuggingFaceH4/rlaif-v_formatted \
>     --model_name_or_path google/paligemma-3b-pt-224 \
>     --trust_remote_code \
>     --per_device_train_batch_size 1 \
>     --gradient_accumulation_steps 8 \
>     --output_dir dpo_paligemma_rlaif-v \
>     --bf16 \
>     --torch_dtype bfloat16
> ```

I run the same code for llava-hf/llava-1.5-7b-hf, but it says:
`'LlavaProcessor' object has no attribute 'apply_chat_template'` I tried transformers==4.37.2 and 4.36.2 but failed. what's your transformers version?

**Comment by Liuziyu77** [comment]
> Closes #1784
> 
> Llava:
> 
> ```shell
> accelerate launch examples/scripts/dpo_visual.py \
>     --dataset_name HuggingFaceH4/rlaif-v_formatted \
>     --model_name_or_path llava-hf/llava-1.5-7b-hf \
>     --per_device_train_batch_size 1 \
>     --gradient_accumulation_steps 8 \
>     --output_dir dpo_llava_rlaif-v \
>     --bf16 \
>     --torch_dtype bfloat16 \
>     --use_peft \
>     --lora_target_modules=all-linear
> ```
> 
> PaliGemma:
> 
> ```shell
> accelerate launch examples/scripts/dpo_visual.py \
>     --dataset_name HuggingFaceH4/rlaif-v_formatted \
>     --model_name_or_path google/paligemma-3b-pt-224 \
>     --trust_remote_code \
>     --per_device_train_batch_size 1 \
>     --gradient_accumulation_steps 8 \
>     --output_dir dpo_paligemma_rlaif-v \
>     --bf16 \
>     --torch_dtype bfloat16
> ```

If I want to support multi-pics DPO, where can I modify the code? Thank you for your patient very much!

**Comment by qgallouedec** [comment]
@Liuziyu77 do you mind opening an issue for your questions? It's way easier for us to keep track of them

**Comment by g0kul6** [comment]
Could we possibly try llava_next_vicuna_7b ?

**Comment by qgallouedec** [comment]
You can give it a try. Feel free to open an issue, it will help us tracking progress
